bigframes/core/compile/polars/compiler.py

```diff
@@ -0,0 +1,373 @@
+else:
+    try:
+        import polars as pl
+    except Exception:
+        polars_installed = False
+
+if polars_installed:
+
+    @dataclasses.dataclass(frozen=True)
+    class PolarsExpressionCompiler:
+        """
+        Simple compiler for converting bigframes expressions to polars expressions.
+
+        Should be extended to dispatch based on bigframes schema types.
+        """
+
+        @functools.singledispatchmethod
+        def compile_expression(self, expression: ex.Expression):
+            raise NotImplementedError(f"Cannot compile expression: {expression}")
+
+        @compile_expression.register
+        def _(
+            self,
+            expression: ex.ScalarConstantExpression,
+        ):
+            return pl.lit(expression.value)
+
+        @compile_expression.register
+        def _(
+            self,
+            expression: ex.DerefOp,
+        ):
+            return pl.col(expression.id.sql)
+
+        @compile_expression.register
+        def _(
+            self,
+            expression: ex.OpExpression,
+        ):
+            # TODO: Complete the implementation, convert to hash dispatch
+            op = expression.op
+            args = tuple(map(self.compile_expression, expression.inputs))
+            if isinstance(op, ops.invert_op.__class__):
+                return args[0].neg()
+            if isinstance(op, ops.and_op.__class__):
+                return args[0] & args[1]
+            if isinstance(op, ops.or_op.__class__):
+                return args[0] | args[1]
+            if isinstance(op, ops.add_op.__class__):
+                return args[0] + args[1]
+            if isinstance(op, ops.sub_op.__class__):
+                return args[0] - args[1]
+            if isinstance(op, ops.ge_op.__class__):
+                return args[0] >= args[1]
+            if isinstance(op, ops.gt_op.__class__):
+                return args[0] > args[1]
+            if isinstance(op, ops.le_op.__class__):
+                return args[0] <= args[1]
+            if isinstance(op, ops.lt_op.__class__):
+                return args[0] < args[1]
+            if isinstance(op, ops.eq_op.__class__):
+                return args[0] == args[1]
+            if isinstance(op, ops.mod_op.__class__):
+                return args[0] % args[1]
+            if isinstance(op, ops.coalesce_op.__class__):
+                return pl.coalesce(*args)
+            if isinstance(op, ops.CaseWhenOp):
+                expr = pl.when(args[0]).then(args[1])
+                for pred, result in zip(args[2::2], args[3::2]):
+                    return expr.when(pred).then(result)
+                return expr
+            raise NotImplementedError(f"Polars compiler hasn't implemented {op}")
+
+    @dataclasses.dataclass(frozen=True)
+    class PolarsAggregateCompiler:
+        scalar_compiler = PolarsExpressionCompiler()
+
+        def get_args(
+            self,
+            agg: ex.Aggregation,
+        ) -> Sequence[pl.Expr]:
```

Comment:
Could we add a doc string here explaining the need for it (projection-y stuff, IIRC?)